botocore/parsers.py

```diff
@@ -694,10 +695,24 @@ def _do_error_parse(self, response, shape):
         error['Error']['Message'] = body.get(
             'message', body.get('Message', '')
         )
+        # Error response may contain an x-amzn-query-error header for json
+        # we need to fetch the error code from this header in that case
+        query_error = header.get('x-amzn-query-error') if header else None
+        query_error_code = None
+        if (
+            query_error
+            and query_error.split(';')
+            and len(query_error.split(';')) == 2
+        ):
+            query_error_code = query_error.split(';')[0]
         # if the message did not contain an error code
         # include the response status code
         response_code = response.get('status_code')
-        code = body.get('__type', response_code and str(response_code))
+        code = (
+            query_error_code
+            if query_error_code
+            else body.get('__type', response_code and str(response_code))
+        )
```

Comment:
We'll need to move the block above here so we have access to the response code. That will keep the logical grouping tighter as well.

Reply:
Sounds good. Will move logic block together.